File: ios/Sources/ARYPush/Internal/RetryManager.swift
import Foundation

/// Exponential backoff with jitter.
///
/// Jitter is not decoration. Without it, every device that failed against the same backend
/// outage retries at the same instant and re-creates the outage. The SDK is embedded in many
/// applications, so a synchronised retry storm is a realistic failure mode rather than a
/// theoretical one.
struct RetryManager {

    private let config: RetryConfig
    private let randomProvider: () -> Double

    init(config: RetryConfig, randomProvider: @escaping () -> Double = { Double.random(in: 0..<1) }) {
        self.config = config
        self.randomProvider = randomProvider
    }

    var maxAttempts: Int { config.maxAttempts }

    /// True while another attempt is permitted. `attempt` is 1-based.
    func canRetry(attempt: Int) -> Bool { attempt < config.maxAttempts }

    /// Delay before the attempt following `attempt`.
    ///
    /// A server-supplied `Retry-After` always wins when present and permitted, clamped so that a
    /// hostile or mistaken header cannot park the queue indefinitely.
    func delay(attempt: Int, retryAfter: TimeInterval? = nil) -> TimeInterval {
        if config.respectRetryAfter, let retryAfter, retryAfter > 0 {
            return min(retryAfter, config.maxRetryAfter)
        }

        let exponent = Double(max(0, attempt - 1))
        let exponential = config.initialBackoff * pow(config.backoffMultiplier, exponent)
        let capped = min(exponential, config.maxBackoff)
        guard config.jitterFactor > 0 else { return capped }

        // Full jitter at factor 1.0: uniform over [0, capped]. Lower factors keep a fixed floor
        // so backoff still grows predictably.
        let floor = capped * (1 - config.jitterFactor)
        return max(0, floor + randomProvider() * (capped - floor))
    }
}

File: android/sdk/src/main/kotlin/com/ary/push/internal/sync/RetryManager.kt
package com.ary.push.internal.sync

import com.ary.push.RetryConfig
import kotlin.math.min
import kotlin.math.pow
import kotlin.random.Random

/**
 * Exponential backoff with jitter.
 *
 * Jitter is not decoration. Without it, every device that failed against the same backend
 * outage retries at the same instant and re-creates the outage; the SDK is embedded in many
 * applications, so a synchronised retry storm is a realistic failure mode rather than a
 * theoretical one.
 */
internal class RetryManager(
    private val config: RetryConfig,
    private val random: Random = Random.Default
) {

    /** True while another attempt is permitted. [attempt] is 1-based. */
    fun canRetry(attempt: Int): Boolean = attempt < config.maxAttempts

    /**
     * Delay before the attempt following [attempt].
     *
     * A server-supplied `Retry-After` always wins when present and permitted, clamped so that a
     * hostile or mistaken header cannot park the queue indefinitely.
     */
    fun delayMillis(attempt: Int, retryAfterMs: Long? = null): Long {
        if (config.respectRetryAfter && retryAfterMs != null && retryAfterMs > 0) {
            return min(retryAfterMs, config.maxRetryAfterMs)
        }
        val exponential = config.initialBackoffMs *
            config.backoffMultiplier.pow((attempt - 1).coerceAtLeast(0))
        val capped = min(exponential, config.maxBackoffMs.toDouble())
        if (config.jitterFactor <= 0.0) return capped.toLong()

        // Full jitter at factor 1.0: uniform over [0, capped]. Lower factors keep a fixed
        // floor so that backoff still grows predictably.
        val floor = capped * (1.0 - config.jitterFactor)
        val span = capped - floor
        return (floor + random.nextDouble() * span).toLong().coerceAtLeast(0L)
    }

    /** Total attempts permitted, including the first. */
    val maxAttempts: Int get() = config.maxAttempts
}
